src/pages/dataElements/form/DataElementForm.tsx

```diff
@@ -0,0 +1,286 @@
+import i18n from '@dhis2/d2-i18n'
+import {
+    InputFieldFF,
+    SingleSelectFieldFF,
+    TextAreaFieldFF,
+} from '@dhis2/ui'
+import React, { useState } from 'react'
+import { Field as FieldRFF } from 'react-final-form'
+import {
+    StandardFormSection,
+    StandardFormSectionTitle,
+    StandardFormSectionDescription,
+    StandardFormField,
+} from '../../../components'
+import {
+    ColorAndIconField,
+    DomainField,
+    LegendsField,
+} from './custom-fields'
+import classes from './DataElementForm.module.css'
+import { EditableFieldWrapper } from './EditableFieldWrapper'
+import {
+    useCategoryCombosQuery,
+    useOptionSetsQuery,
+    useOptionSetCommentsQuery,
+    useLegendsQuery,
+    useAddLegendMutation,
+    useAddCategoryComboMutation,
+    useAddOptionSetMutation,
+    useAddOptionSetCommentMutation,
+} from './hooks'
+
+export function DataElementForm() {
+    const categoryCombos = useCategoryCombosQuery()
+    const optionSets = useOptionSetsQuery()
+    const optionSetComments = useOptionSetCommentsQuery()
+    const legends = useLegendsQuery()
+
+    const [addingLegend, setAddingLegend] = useState(false)
+    const [addingCategoryCombo, setAddingCategoryCombo] = useState(false)
+    const [addingOptionSet, setAddingOptionSet] = useState(false)
+    const [addingOptionSetComment, setAddingOptionSetComment] = useState(false)
+
+    // @TODO(DataElementForm): Use these
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [addLegend] = useAddLegendMutation()
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [addCategoryCombo] = useAddCategoryComboMutation()
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [addOptionSet] = useAddOptionSetMutation()
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [addOptionSetComment] = useAddOptionSetCommentMutation()
```

Comment:
I'm not sure if these should be part of the form? This will undoubtedly be shared, and I think we could extract these into components, that wraps this logic there.

Eg. `LegendField`.

Reply:
> In general I want the Forms themselves to have as little logic as possible, so it's easy to compose these fields in other forms.

Yeah, I agree with that. I only put these here because I didn't have a better place where I could put them. I just wanted something to be there and while it shouldn't be too much work to move code around.

Right now I'm not sure what should happen at all with these special form fields that have the two buttons. If you could open a modal and add a new element, then the "refresh" button makes little sense to me as we can automatically refetch items after having added a new one.

So maybe clicking on the "add" button will simply open a new tab with the actual form? Maybe we'll have some "initial data" parameters, e.g. when adding new attributes, it will automatically have "data elements" selected, but that's just a thought. But opening a new tab seems like the simplest solution implementation-wise as we wouldn't have to render the entire form in a Modal as well, but provide the full user experience with the real form on a full-width page. In that context the "refresh" button makes much more sense.

Otherwise we'll either have to implement two forms for each metadata that can be added via modals as well (e.g. we don't want to show the "add" button in a form inside a modal) or we'll have to have control flow changes / conditions in the forms that will produce a different form UI, but that'd make forms much more complicated compared to a "stupid" form that renders just fields.

In case we want to have re-usable components for adding certain metadata, e.g. the category combo dropdown with the buttons, I agree that it'd make sense to have them in their own component and move the hooks there instead of in the form.

> I think we might also need to have "model"-specific fields, that would wrap "field-rule"-logic.

I was thinking about that, too. The solution I had in mind was to have a hook that uses RFF's `useFormSpy` internally and modifies form state reactively. This way, we'd separate logic that combines two fields without having to mix that logic with the fields themselves.

In the beginning I'd make these super simple and implement the logic for which field change would trigger another effect directly. I'm sure that after some time - when we see a pattern emerge - we can even make these hooks more abstract have something that we can re-use across forms without having to duplicate stupid code everywhere

Reply:
> Yeah, I agree with that. I only put these here because I didn't have a better place where I could put them. I just wanted something to be there and while it shouldn't be too much work to move code around.

Sounds good.

> But opening a new tab seems like the simplest solution implementation-wise as we wouldn't have to render the entire form in a Modal as well, but provide the full user experience with the real form on a full-width page.

Yeah, agree. Let's talk to Joe about it. If a form just composed simple fields, it wouldn't necessarily be that hard to implement a modal version of it - or as you say have some "control" props that conditionally render things, if needed. But I agree we can defer that.


> I was thinking about that, too. The solution I had in mind was to have a hook that uses RFF's useFormSpy internally and modifies form state reactively. This way, we'd separate logic that combines two fields without having to mix that logic with the fields themselves.

Sounds that we're on the same page, and I've been thinking about something similar. Not sure if we would even need a custom hook at first - but would definitely be great if we could find a fitting API that was generic enough. 
One potential challenge that I can anticipate, is for the "dependant" field to get access to another fields data. Since naively, eg the optionSet data in the form would just be the ID of the selected optionSet - but the rule would need the `valueType` of that optionSet as well. So it's something to think about - maybe we can pass the entire object as value, and use `format` and `parse`.

Also, `FormSpy` doesn't support subscribing to a single field - you would have to subscribe to all values. We might be able to use `useField` in "dependant" components - but not really sure if that's a good idea (eg. DataElementValueType would call `useField('optionSet')`).
So we might indeed need to subscribe to `values`, and check for the condition in `onChange` and update values imperatively in the onChange-handler. 
Could look at https://github.com/final-form/final-form-calculate for inspiration. 
However another thing that complicates this, is that rules doesn't just affect the `value` of other fields - it also affects eg. `hidden`-functionality and other render-states like `disabled`. So to me it seems it would be better to do these subscriptions 'reactively' like you say - eg. use `useField('optionSet')` (if possible). And then an `useEffect` to update the "current"-fields value (eg. DataElementValueType).

However these limitations, and the amount of unresolved issues (and serious issues like this: https://github.com/final-form/react-final-form/issues/984 )  for `react final form`, was why I initially planned to test out `react-hook-form`.